### PR TITLE
Add support for multi fields relations

### LINKED ISSUE #128: Add management of multi fields relation
Actually it is possible to work with relations that have just **one** field in relation between two layers.
Start to handle multi fields relation will be necessary

### PR DESCRIPTION
**Pull request by volterra79**
Close: https://github.com/g3w-suite/g3w-client/issues/128

Related to:

- https://github.com/g3w-suite/g3w-admin/pull/596
- https://github.com/g3w-suite/g3w-client-plugin-editing/pull/53

## • father_relation

image

## • child_relation

image

## • father_relation / child_relation

image

## QGIS Configuration ---> [9.3.8. Relations Properties](https://docs.qgis.org/3.22/en/docs/user_manual/introduction/qgis_configuration.html?highlight=relation#relations-properties)

image

image

### CODE REVIEW

File: src/app/core/relations/relation.js [hunk 1]
```diff
@@ -1,98 +1,154 @@
...
   this.state.name = name;
 };
 
+/**
+ * @TODO check if this is currently a deprecated feature
+ * 
+ * @returns {string} relation title
+ */
 proto.getTitle = function() {
   return this.state.title;
```
Comment (Raruto) [inline src/app/core/relations/relation.js:81]:
```js
this.state.title
```

it is not defined in the class's constructor, check if this is a refactoring oversight

File: src/app/core/relations/relation.js [hunk 2]
```diff
@@ -1,98 +1,154 @@
...
 };
 
+/**
+ * @TODO check if this is currently a deprecated feature
+ * 
+ * @param {string} title relation title
+ * @returns {void}
+ */
 proto.setTitle = function(title) {
   return this.state.title = title;
```
Comment (Raruto) [inline src/app/core/relations/relation.js:94]:
```js
this.state.title
```

it is not defined in the class's constructor, check if this is a refactoring oversight

File: src/app/core/relations/relation.js [hunk 3]
```diff
@@ -1,98 +1,154 @@
...
 
+/**
+ * @returns {object} relation fields
+ */
 proto.getFields = function() {
   return {
-    father: this.state.fatherField,
-    child: this.state.childField
+    father: this.state.fatherFields, // temp
+    child: this.state.childFields // temp
```
Comment (Raruto) [inline src/app/core/relations/relation.js]:
```js
// temp
```

improve documentation about it (is this just a typo? is it a variable that was temporarily created and then immediately deleted? does it contain a temporarily generated value?)

File: src/app/core/relations/relations.js [hunk 4]
```diff
@@ -111,32 +181,58 @@ proto.removeRelation = function(relation) {
...
-// get children based on father id
+/**
+ * Get children based on father id
+ * 
+ * @param {string} fatherId relation identifier
+ * @returns {object} children
+ */
 proto.getChildren = function(fatherId) {
   if (!this.isFather(fatherId)) return null;
   return this._relationsInfo.fathers[fatherId];
```
Comment (Raruto) [inline src/app/core/relations/relations.js]: Here the code is not very clear (there seems to be a mistake), why calling the following function:

```js
Relations::getChildren(fatherId)
```

is the father returned?

```js
return this._relationsInfo.fathers[fatherId];
```

File: src/app/core/relations/relations.js [hunk 5]
```diff
@@ -111,32 +181,58 @@ proto.removeRelation = function(relation) {
...
-// get fathers based on childId
+/**
+ * Get fathers based on childId
+ * 
+ * @prop {string} childId relation identifier
+ * @returns {object} fathers
+ */
 proto.getFathers = function(childId) {
   if (!this.isChild(childId)) return null;
   return this._relationsInfo.children[childId];
```
Comment (Raruto) [inline src/app/core/relations/relations.js]: Here the code is not very clear (there seems to be a mistake), why calling the following function:

```js
Relations::getFathers(childId)
```

is the children returned?

```js
return this._relationsInfo.children[childId];
```

File: src/app/core/relations/relations.js [hunk 6]
```diff
@@ -2,141 +2,237 @@ const { base, inherit } = require('core/utils/utils');
...
+ * @private
+ */
 proto._clearRelationsInfo = function() {
   this._relationsInfo = {
-    children: {},
-    fathers: {},
-    father_children: {}
+    children:     {},     // array child (unique ids)
+    fathers:      {},     // array father (unique ids)
+    father_child: {},     // info parent child
```
Comment (Raruto) [inline src/app/core/relations/relations.js]: @volterra79 double check this assignment.

it should be `father_children` or `father_child`?

**Comment by Raruto** [comment]
Tested with: [editing-multi-fields.qgs](https://github.com/g3w-suite/g3w-suite-demo-projects/commit/b7ca4bca036d5ba9112a5c994bbcd84525176114)

File: src/app/core/relations/relation.js (hunk 3 above)
Comment (volterra79) [reply src/app/core/relations/relation.js]: @Raruto outdate changes

File: src/app/core/relations/relations.js (hunk 4 above)
Comment (volterra79) [reply src/app/core/relations/relations.js]: @Raruto this._relationsInfo.fathers store layers that are father in relations and store layerId children

File: src/app/core/relations/relations.js (hunk 5 above)
Comment (volterra79) [reply src/app/core/relations/relations.js]: @Raruto  this._relationsInfo.children store layers that are child in relations and store layerId fathers

File: src/app/core/relations/relations.js (hunk 6 above)
Comment (volterra79) [reply src/app/core/relations/relations.js]: @Raruto it is correct father_child. It store father layerId and child layerId as attribute and relationId as value

**Review by Raruto — APPROVED** [review]

**Pull request merged**